# Organ-Specific Microbiomes of Biomphalaria Snails

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.15122183.svg)](https://doi.org/10.5281/zenodo.15122183) 

## Abstract
The microbiome is increasingly recognized to shape many aspects of its host biology and is a key determinant of health and disease. The microbiome may influence transmission of pathogens by their vectors, such as mosquitoes or aquatic snails. We previously sequenced the bacterial 16S V4 ribosomal DNA of the hemolymph (blood) of *Biomphalaria* spp. snails, one of the vectors of the human blood fluke schistosome. We showed that snail hemolymph harbored an abundant and diverse microbiome. This microbiome is distinct from the water environment and can discriminate snail species and populations. As hemolymph bathes snail organs, we then investigated the heterogeneity of the microbiome in these organs. We dissected ten snails for each of two different species (*B. alexandrina* and *B. glabrata*) and collected their organs (ovotestis, hepatopancreas, gut, and stomach). We also ground in liquid nitrogen four whole snails of each species. We sampled the water in which the snails were living (environmental controls). Sequencing the 16S V4 rDNA revealed organ-specific microbiomes. These microbiomes harbored a lower diversity than the hemolymph microbiome, and the whole-snail microbiome. The organ microbiomes tend to cluster by physiological function. In addition, we showed that the whole-snail microbiome is more similar to hemolymph microbiome. These results are critical for future work on snail microbiomes and show the necessity of sampling individual organ microbiomes to provide a complete description of snail microbiomes.

## Code

The code used to analyze the data presented in the mansucript is available in this repository.

## Prerequisites

Two dependencies must be installed before running the Jupyter notebook:
* A [conda](https://docs.conda.io/en/latest/) distribution, like [miniconda](https://docs.conda.io/en/latest/miniconda.html),
* [Jupyter Notebook](https://jupyter.readthedocs.io/en/latest/install.html) with a [bash kernel](https://github.com/takluyver/bash_kernel).

Once this is done, run the first cells of the Jupyter notebook from this repository to install a dedicated conda environment and the required R packages.
